cloudinit/net/network_state.py

```diff
@@ -1035,24 +1037,7 @@ def ipv4_mask_to_net_prefix(mask):
        str(24)         => 24
        "24"            => 24
     """
-    if isinstance(mask, int):
-        return mask
-    if isinstance(mask, str):
-        try:
-            return int(mask)
-        except ValueError:
-            pass
-    else:
-        raise TypeError("mask '%s' is not a string or int")
-
-    if "." not in mask:
-        raise ValueError("netmask '%s' does not contain a '.'" % mask)
-
-    toks = mask.split(".")
-    if len(toks) != 4:
-        raise ValueError("netmask '%s' had only %d parts" % (mask, len(toks)))
-
-    return sum([bin(int(x)).count("1") for x in toks])
+    return mask_to_net_prefix(mask, 32)
```

Comment:
It's probably worth pointing out that the standard library implements the functionality required for `ipv4_mask_to_net_prefix()` already.

This could easily be:
```
def ipv4_mask_to_net_prefix(mask):
    return ipaddress.IPv4Network(f"0.0.0.0/{mask}").prefixlen
```

Reply:
Yes, I'm aware of that. We can of course opt to use the functionality from the standard library in case the caller uses the ip version specific method.

My reasoning was that in cloud-init there are many places where mask_to_net_prefix() is called directly without the caller's knowledge if it is IPv6 or IPv4. I figured since mask_to_net_prefix() here works for both IPv4 and IPv6 these version specific methods can just as well use mask_to_net_prefix.

If you feel strongly about using the library, in  ipv4_mask_to_net_prefix I can change it.

Reply:
> > NetworkManager will ignore the routes, as it sees the entry as invalid when the netmask is in expanded format.
> 
> I did a little bit of digging. This format is not valid a valid netmask notation for ipv6, hence the lack of support in openstack and python standard library. I need to do a little bit of digging to better understand the rendering code, but if this format is coming from netplan we probably need to push on netplan to use a valid format.
> 

++ from me on changeing the source, but afaik this coming from opentack's network metadata.
See: https://docs.openstack.org/nova/latest/user/metadata.html#openstack-format-metadata
Their Schema: https://docs.openstack.org/nova/latest/_downloads/9119ca7ac90aa2990e762c08baea3a36/network_data.json


To quote the network_data.json schema:
```
    "l3_ipv6_netmask": {
      "$id": "#/definitions/l3_ipv6_netmask",
      "type": "string",
      "pattern": "^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7})|(::))$",
      "title": "L3 IPv6 network mask",
      "examples": [
        "ffff:ffff:ffff:ffff::"
      ]
    },
```


Openstack Nova is using 'netaddr', and it can do this:
```
>>> route_addr = netaddr.IPNetwork('2bd0::/63')
>>> route_addr
IPNetwork('2bd0::/63')
>>> route_addr.netmask
IPAddress('ffff:ffff:ffff:fffe::')
```

Here is the Nova code building the network_metadata, it shared code for IPv4 and IPv6.
https://opendev.org/openstack/nova/src/branch/master/nova/virt/netutils.py#L311-L319